src/Walker/Variants/WalkerQueue.cpp

```diff
@@ -9,11 +9,13 @@ Path WalkerQueue::Walk(BaseHeuristic *heuristic, const PDDL::State state) {
 
     auto actions = actionGenerator.GenerateActions(&currentState);
     for (auto iter = actions.begin(); iter != actions.end(); iter++) {
+        if (!widthFunc->Iterate(current))
+            break;
         currentPath.steps.push_back((*iter));
-        auto changes = currentState.DoAction(&(*iter));
-        auto eval = heuristic->Eval(&currentState);
-        searchQueue.emplace(std::make_pair(eval, std::make_pair(currentState, currentPath)));
-        currentState.UndoAction(&changes);
+        auto eval = heuristic->Eval(&currentState, &(*iter));
+        PDDL::State newState = currentState;
+        newState.DoAction(&(*iter));
+        searchQueue.emplace(std::make_pair(eval, std::make_pair(newState, currentPath)));
         currentPath.steps.pop_back();
```

Comment:
```suggestion
    auto changes = currentState.DoAction(&(*iter));
        auto eval = heuristic->Eval(&currentState, nullptr);
        searchQueue.emplace(std::make_pair(eval, std::make_pair(currentState, currentPath)));
        currentState.UndoAction(&changes);
        currentPath.steps.pop_back();
```
There is no reason to make a duplicate state twice.